Implement filesystem abstraction

joshuagl opened the pull request:
**Note**: this branch of securesystemslib doesn't currently work with theupdateframework/tuf – it's very much WIP.

It feels like the shape these changes are taking introduces a conflict between the desire to support abstract filesystem backends vs. reasonable feature enhancements that have been suggested such as #222. It's not yet clear to me how we will support both abstract filesystem backends _and_ creating private key files with restricted privileges.

**Fixes issue #**: step one, two and four of theupdateframework/tuf#1009

**Description of the changes being introduced by the pull request**:
* Add an abstract base class defining an interface for storage
* Add a concrete implementation of the storage interface for accessing the local file system 
* Add abstract storage support to `securesystemslib.hash.digest_filename()`
* Add abstract storage support to `securesystemslib.util.[get_file_details|ensure_parent_dir|load_json_file]()`

**Please verify and check that the pull request fulfils the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

mnm678 commented on securesystemslib/storage.py:
```diff
@@ -0,0 +1,219 @@
+
+
+
+class FilesystemBackend(StorageBackendInterface):
```
Should this be FileSystemBackend?

joshuagl replied:
I've seen file system and filesystem used, I don't have a strong preference. I'd be happy to change it.

woodruffw replied:
I have a _slight_ preference for `FilesystemBackend` based on "filesystem" being a full noun, but this backend doesn't affect me in either case 😉

mnm678 commented on securesystemslib/util.py:
```diff
@@ -77,15 +89,15 @@ def get_file_details(filepath, hash_algorithms=['sha256']):
     raise securesystemslib.exceptions.Error('Path ' + repr(filepath) + ' doest'
         ' not exist.')
 
-  filepath = os.path.abspath(filepath)
```
Is this handled by the FilesystemBackend?

joshuagl replied:
It isn't handled by the FilesystemBackend. This would probably be OK to leave in, as `abspath()` doesn't access the underlying files or directories.

We don't do this kind of normalisation everywhere we take paths, thus my thought process when making this change was to try and move towards expecting normalised paths in securesystemslib functions. That's probably better as a separate PR and discussion, though.

I'll force push a change to add this back.

joshuagl commented:
Thanks for the review @mnm678

> This looks good to me, but it would be nice to get a review from a non-TUF ssl user as well.

Agreed. @SantiagoTorres, would you be able to take a look?

> Regarding #222, would it be possible to require this as part of the abstract class? There could be a put_secure method that uses the permissions you would want for a key file.

That's an interesting suggestion. @sechkova proposed similar when we discussed #222 offline. 
What I don't know is, if we added a `put_secure()` or similar to `StorageBackendInterface`, would that make sense on systems like S3 and GCS? 
How would it be implemented? Would they store the permission bits as metadata on the file and apply it during `get()`?

Perhaps @woodruffw can help us understand how this might work on S3 and similar storage?

joshuagl commented:
I just tested this branch with the _develop_ branch of in-toto and the tests run without issue.

woodruffw commented on securesystemslib/interface.py:
```diff
@@ -664,10 +694,11 @@ def import_ed25519_privatekey_from_file(filepath, password=None, prompt=False):
 
   # Finally, regardless of password, try decrypting the key, if necessary.
   # Otherwise, load it straight from the disk.
```
Nit: `the disk` -> `storage`?

woodruffw commented on securesystemslib/storage.py:
```diff
@@ -0,0 +1,219 @@
+    local filesystems using Python standard library functions.
+  """
+
+  def get(self, filepath):
```
Proposal: Can we turn this into a context manager? Right now, this requires the following API pattern:

```python
foo = backend.get("foo")
# operations
foo.close()
```

...which makes it easy for a consumer to accidentally leak a file descriptor. With [contextlib](https://docs.python.org/3/library/contextlib.html), we could do something like this:

```python
with backend.get("foo") as foo:
  # operations...
```

joshuagl replied:
Great suggestion, thanks. I want to spend a bit of time with the documentation to figure out a clean way to do this without _requiring_ use of a context manager. I would like to be able to support the current explicit `get()`/`close()` pattern as well as the `with get() as ...` pattern.

Would you expect the interface to indicate/recommend implementing a context manager? When I looked during initial work on this PR I couldn't see an obvious way to do that beyond documentation.

joshuagl replied:
I've converted `get()` to a ContextManager in both the interface (docstrings) and the `FilesystemBackend`. As implemented `get()` can only be used as a context manager. I'd appreciate any thoughts on that.

joshuagl commented:
I've updated the PR here to note the `get()` function of the interface should be a context manager and updated the _FilesystemBackend_ implementation to return a ContextManager from `get()`. 

I've also addressed other comments from @mnm678 and @woodruffw. I'd very much appreciate some re-review.

woodruffw commented on securesystemslib/storage.py:
```diff
@@ -0,0 +1,248 @@
+  """
+
+  class GetFile(object):
+    # Implementing get() as a function with the @contextmanager decorator
```
Thanks for this explanation!

woodruffw commented on securesystemslib/storage.py:
```diff
@@ -0,0 +1,248 @@
+
+
+
+# Map our class ContextManager implementation to the function expected of the
```
Nit: Mismatched indentation? Ignore me if it's a style choice 🙂

joshuagl replied:
Thanks for catching that. It wasn't intentional. Fixed in a force-push
